### PR TITLE
Implement traits system

### PR DESCRIPTION
**Pull request by Visne**
## About the PR
This PR adds a traits system and a smoker trait.

The traits system allows the player to pick traits in the character editor for RP reasons or extra challenge. Regarding code, the traits system essentially consists of an EntitySystem that applies components to a player when they join. The components that will be added can be specified in a YAML trait prototype:

```cs
- type: trait
  id: SmokerTrait
  name: Smoker
  components:
  - type: SmokerTrait
```

This then adds `SmokerTraitComponent` to the player at roundstart, and at this point `SmokerTraitSystem` takes over to handle the actual logic of the smoker trait.

~~The smoker trait is very similar to hunger and thirst. After 6 minutes, the player gets an alert that they are feeling light nicotine cravings. After 10 minutes, they experience intense nicotine cravings and lose 10% of their movement speed. The debuff is small, to encourage people to play with the trait and not disadvantage players that care a lot about RP too much.~~

~~The code is essentially done, and can be reviewed already. Only two more small things need to be done:~~
- [x] ~~Replace codersprites with real sprites (volunteers welcome)~~
- [x] ~~Make smokables take a fixed amount of time to smoke, as opposed to depending on how much solution is stored in them~~

~~Changelog~~
- ~~add: Traits! You can pick them in the character editor for extra roleplay or extra challenge.~~
- ~~add: Added the smoker trait. If you enable it, you have to smoke once in a while or experience a small movement speed debuff.~~

EDIT: I've removed the smoker trait from this PR, so this PR just adds the traits system. After this is merged, the players will see a traits tab with "No traits available" in it.

### CODE REVIEW

File: Content.Server/Nutrition/EntitySystems/ThirstSystem.cs [hunk 1]
```diff
@@ -43,9 +38,6 @@ private void OnComponentStartup(EntityUid uid, ThirstComponent component, Compon
 
         private void OnRefreshMovespeed(EntityUid uid, ThirstComponent component, RefreshMovementSpeedModifiersEvent args)
         {
-            if (_jetpack.IsUserFlying(component.Owner))
```
Comment (Elijahrane) [inline Content.Server/Nutrition/EntitySystems/ThirstSystem.cs:46]: what does this have to do with traits

Comment (Visne) [reply Content.Server/Nutrition/EntitySystems/ThirstSystem.cs:46]: Nothing, but I came across it and it's not supposed to be there. I could make it a separate PR if you want.

**Comment by Visne** [comment]
Ready for review!

**Review by PJB3005 — APPROVED** [review]
DB code looks fine

File: Content.Server/Chemistry/ReagentEffects/SatiateNicotineCraving.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    /// <summary>
+    /// Satiate craving if player has <see cref="SmokerTraitComponent"/>.
+    /// </summary>
+    public override void Effect(ReagentEffectArgs args)
+    {
+        if (!args.EntityManager.TryGetComponent(args.SolutionEntity, out SmokerTraitComponent? trait))
+            return;
+
+        var smokerTraitSystem = IoCManager.Resolve<IEntitySystemManager>().GetEntitySystem<SmokerTraitSystem>();
```
Comment (metalgearsloth) [inline Content.Server/Chemistry/ReagentEffects/SatiateNicotineCraving.cs:25]: You can get IEntitySystemManager from EntityManager

File: Content.Server/Nutrition/EntitySystems/ThirstSystem.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Server/Nutrition/EntitySystems/ThirstSystem.cs:46]: Is there corresponding code to fix the mispredict?

File: Content.Server/Traits/TraitSystem.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+using Content.Shared.Traits;
+using Robust.Shared.Prototypes;
+using Robust.Shared.Serialization.Manager;
+
+namespace Content.Server.Traits;
+
+public sealed class TraitSystem : EntitySystem
+{
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+    [Dependency] private readonly IEntityManager _entityManager = default!;
```
Comment (metalgearsloth) [inline Content.Server/Traits/TraitSystem.cs:11]: This is already on entitysystems as a dependency.

File: Content.Shared/Traits/TraitPrototype.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    /// </summary>
+    [Prototype("trait")]
+    public sealed class TraitPrototype : IPrototype
+    {
+        [ViewVariables]
+        [IdDataField]
+        public string ID { get; } = default!;
+
+        /// <summary>
+        ///     The name of this antag as displayed to players.
```
Comment (metalgearsloth) [inline Content.Shared/Traits/TraitPrototype.cs:17]: Comments need updating.

Comment (Visne) [reply Content.Shared/Traits/TraitPrototype.cs:17]: Oops

File: Content.Server/Nutrition/EntitySystems/ThirstSystem.cs (hunk 1 above)
Comment (Visne) [reply Content.Server/Nutrition/EntitySystems/ThirstSystem.cs:46]: Undid the changes for now, and added a TODO

**Comment by Visne** [comment]
Can be reviewed again

**Review by mirrorcult — CHANGES_REQUESTED** [review]
is there any reason the specific `NicotineCraving`/`Smoker` stuff can't just be generalized into an addiction system that works for any reagent? most of it is already pretty generic, you'd basically just put everything in a dict on a component instead. and the trait already serializes a component registry, so you could make addictions for various chemicals. i don't think this would be particularly hard to do

**Comment by Visne** [comment]
Smoker trait is now removed from this PR, and it just adds the traits system. Addiction/needs will be a separate PR.

**Review by mirrorcult — APPROVED** [review]
